cmd/convertor/builder/builder_utils.go

```diff
@@ -22,52 +22,99 @@ import (
+func fetch(ctx context.Context, fetcher remotes.Fetcher, desc specs.Descriptor, target any) error {
+	rc, err := fetcher.Fetch(ctx, desc)
 	if err != nil {
-		return specs.Manifest{}, specs.Image{}, errors.Wrapf(err, "failed to fetch manifest")
+		return fmt.Errorf("failed to fetch digest %v: %w", desc.Digest, err)
 	}
+	defer func() {
+		rc.Close()
+	}()
+
 	buf, err := io.ReadAll(rc)
 	if err != nil {
-		return specs.Manifest{}, specs.Image{}, errors.Wrapf(err, "failed to read manifest")
+		return fmt.Errorf("failed to read digest %v: %w", desc.Digest, err)
 	}
-	rc.Close()
-	manifest := specs.Manifest{}
-	err = json.Unmarshal(buf, &manifest)
-	if err != nil {
-		return specs.Manifest{}, specs.Image{}, err
+	if err = json.Unmarshal(buf, target); err != nil {
+		return fmt.Errorf("failed to unmarshal digest %v: %w", desc.Digest, err)
 	}
-	// get config
-	rc, err = fetcher.Fetch(ctx, manifest.Config)
-	if err != nil {
-		return specs.Manifest{}, specs.Image{}, errors.Wrapf(err, "failed to fetch config")
+	return nil
+}
+
+func fetchManifest(ctx context.Context, fetcher remotes.Fetcher, desc specs.Descriptor) (*specs.Manifest, error) {
+	platformMatcher := platforms.Default()
+	log.G(ctx).Infof("fetching manifest %v with type %v", desc.Digest, desc.MediaType)
+	switch desc.MediaType {
+	case images.MediaTypeDockerSchema2Manifest, specs.MediaTypeImageManifest:
+		config := specs.Manifest{}
+		if err := fetch(ctx, fetcher, desc, &config); err != nil {
+			return nil, fmt.Errorf("failed to fetch manifest: %w", err)
+		}
+		return &config, nil
+	case images.MediaTypeDockerSchema2ManifestList, specs.MediaTypeImageIndex:
+		var target *specs.Descriptor
+		config := specs.Index{}
+		if err := fetch(ctx, fetcher, desc, &config); err != nil {
+			return nil, fmt.Errorf("failed to fetch manifest list: %w", err)
+		}
+		for _, manifest := range config.Manifests {
+			if platformMatcher.Match(*manifest.Platform) {
+				target = &manifest
+				break
+			}
+		}
+		if target == nil {
+			return nil, fmt.Errorf("no match platform found in manifest list")
+		} else {
+			return fetchManifest(ctx, fetcher, *target)
+		}
+	default:
+		return nil, fmt.Errorf("non manifest type digest fetched")
+	}
+}
+
+func fetchConfig(ctx context.Context, fetcher remotes.Fetcher, desc specs.Descriptor) (*specs.Image, error) {
+	config := specs.Image{}
+	if err := fetch(ctx, fetcher, desc, &config); err != nil {
+		return nil, fmt.Errorf("failed to fetch config: %w", err)
 	}
-	buf, err = io.ReadAll(rc)
+	return &config, nil
+}
+
+func fetchManifestAndConfig(ctx context.Context, fetcher remotes.Fetcher, desc specs.Descriptor) (*specs.Manifest, *specs.Image, error) {
+	var manifest *specs.Manifest
+	var config *specs.Image
+	manifest, err := fetchManifest(ctx, fetcher, desc)
 	if err != nil {
-		return specs.Manifest{}, specs.Image{}, errors.Wrapf(err, "failed to read config")
+		return nil, nil, fmt.Errorf("builder: failed to fetch manifest: %w", err)
```

Comment:
Maybe we could rewrite all (or start with the convertor section) `errors.Wrap()` to  `fmt.Errorf(%w)` in a future PR for consistency. Since `github.com/pkg/errors package` is no longer actively maintained.

Reply:
sounds great but that lot of work for sure tho..